pkg/services/ngalert/state/manager.go

```diff
@@ -313,14 +320,22 @@ func (st *Manager) staleResultsHandler(ctx context.Context, alertRule *ngModels.
 			}
 
 			if s.State == eval.Alerting {
-				st.annotateState(ctx, alertRule, s.Labels, time.Now(), eval.Normal, s.State)
+				s.State = eval.Normal
+				s.EndsAt = st.clock.Now()
+				s.Resolved = true
+				s.Stale = true
+				s.LastEvaluationString = ""
+				s.Annotations[ngModels.StateReasonAnnotation] = ngModels.StateReasonMissingSeries
+				st.annotateState(ctx, alertRule, s.Labels, st.clock.Now(), eval.Normal, s.State)
+				resolvedStates = append(resolvedStates, s)
 			}
 		}
 	}
+	return resolvedStates
 }
 
-func isItStale(lastEval time.Time, intervalSeconds int64) bool {
-	return lastEval.Add(2 * time.Duration(intervalSeconds) * time.Second).Before(time.Now())
+func isItStale(timeNow time.Time, lastEval time.Time, intervalSeconds int64) bool {
+	return lastEval.Add(2 * time.Duration(intervalSeconds) * time.Second).Before(timeNow)
```

Comment:
This is the same threshold used by the alertmanager itself to determine if an alert should be expired, right?

In the case where we just drop the alert instance it seems fine to use the same threshold, since there's no communication with the alertmanager system.

But if we use an identical time threshold here in the state manager and in the alertmanager itself, it seems likely we'll have some cases like:
```mermaid
sequenceDiagram

state.Manager ->> alertManager: Instance A Alerting
activate state.Manager
activate alertManager
alertManager -->> alertManager: Instance A Expired
deactivate alertManager
state.Manager ->> alertManager: Instance A Resolved
deactivate state.Manager
```

Should we change either this check or the expiration time to offer ourselves a little leeway?

Reply:
I thought about that. The initial version included the reduction of the stale interval to just 1 eval interval. However, then I reverted it because changing the stale period looked like a much bigger change than I wanted to include. 

The diagram above does not consider the group interval that AM considers while sending resolved messages. The default interval is 5 minutes. In other words, by default, it takes AM up to 5 minutes to send a resolved message.

Now I am thinking about increasing the expiration time so we can make sure that AM does not expire an alert concurrently and Grafana has some time (several seconds) to resolve it.

Reply:
https://github.com/grafana/grafana/blob/afc2b3e438813de81a21b603947506c1a93fdfa5/pkg/services/ngalert/state/state.go#L207-L214

Apparently, auto-resolve in AM happens in at least 3 eval intervals with the bottom interval of 1.5 mins (ResendInterval is hardcoded).

Reply:
I see. As long as we avoid racing resolutions with the alertmanager I'm fine with whatever timing we set up. I broadly agree that explicitly resolving these instances for a given stale label set is a good end-user experience.